ext/node/ops/vm_internal.rs

```diff
@@ -0,0 +1,595 @@
+  };
+  read_only |= attributes.is_read_only();
+
+  if read_only {
+    return;
+  }
+
+  // true for x = 5
+  // false for this.x = 5
+  // false for Object.defineProperty(this, 'foo', ...)
+  // false for vmResult.x = 5 where vmResult = vm.runInContext();
+  let is_contextual_store = ctx.global_proxy(scope) != args.this();
+
+  // Indicator to not return before setting (undeclared) function declarations
+  // on the sandbox in strict mode, i.e. args.ShouldThrowOnError() = true.
+  // True for 'function f() {}', 'this.f = function() {}',
+  // 'var f = function()'.
+  // In effect only for 'function f() {}' because
+  // var f = function(), is_declared = true
+  // this.f = function() {}, is_contextual_store = false.
+  let is_function = value.is_function();
+
+  let is_declared = is_declared_on_global_proxy || is_declared_on_sandbox;
+  if !is_declared
+    && args.should_throw_on_error()
+    && is_contextual_store
+    && !is_function
+  {
+    return;
+  }
+
+  if !is_declared && key.is_symbol() {
+    return;
+  };
+
+  if ctx.sandbox(scope).set(scope, key.into(), value).is_none() {
+    return;
+  }
+
+  if is_declared_on_sandbox {
+    if let Some(desc) =
+      ctx.sandbox(scope).get_own_property_descriptor(scope, key)
+    {
+      if !desc.is_undefined() {
+        let desc_obj: v8::Local<v8::Object> = desc.try_into().unwrap();
+        // We have to specify the return value for any contextual or get/set
+        // property
+        let get_key =
+          v8::String::new_external_onebyte_static(scope, b"get").unwrap();
+        let set_key =
+          v8::String::new_external_onebyte_static(scope, b"get").unwrap();
+        if desc_obj
+          .has_own_property(scope, get_key.into())
+          .unwrap_or(false)
+          || desc_obj
+            .has_own_property(scope, set_key.into())
+            .unwrap_or(false)
+        {
+          rv.set(value);
+        }
+      }
+    }
+  }
+}
+
+fn property_deleter<'s>(
+  scope: &mut v8::HandleScope<'s>,
+  key: v8::Local<'s, v8::Name>,
+  args: v8::PropertyCallbackArguments<'s>,
+  mut rv: v8::ReturnValue,
+) {
+  let ctx = ContextifyContext::get(scope, args.this()).unwrap();
+
+  let context = ctx.context(scope);
+  let sandbox = ctx.sandbox(scope);
+  let context_scope = &mut v8::ContextScope::new(scope, context);
+  if !sandbox.delete(context_scope, key.into()).unwrap_or(false) {
+    return;
+  }
+
+  rv.set_bool(false);
+}
+
+fn property_enumerator<'s>(
+  scope: &mut v8::HandleScope<'s>,
+  args: v8::PropertyCallbackArguments<'s>,
+  mut rv: v8::ReturnValue,
+) {
+  let ctx = ContextifyContext::get(scope, args.this()).unwrap();
+
+  let context = ctx.context(scope);
+  let sandbox = ctx.sandbox(scope);
+  let context_scope = &mut v8::ContextScope::new(scope, context);
+  let Some(properties) = sandbox
+    .get_property_names(context_scope, v8::GetPropertyNamesArgs::default())
+  else {
+    return;
+  };
+
+  rv.set(properties.into());
+}
+
+fn property_definer<'s>(
```

Comment:
Suggestion - since these prop handling methods are talking a bulk of the code, maybe move them to a separate module?

Reply:
It's using internals of `ContextifyContext` and I like that its identical to Node's layout - easier to compare. IMO its better staying in one module unless you feel strongly we should refactor it out.